lib/spack/docs/developer_guide.rst

```diff
@@ -299,6 +315,198 @@ Conceptually, packages are overloaded.  They contain:
 Stage objects
 -------------
 
+
+.. _writing-analyzers:
+
+-----------------
+Writing analyzers
+-----------------
+
+To write an analyzer, you can either include it in one of the files in the
+analyzers module folder (e.g., if it's associated with an application
+binary interface, ABI, you might add it to ``abi.py``) or you can add a new file.
+Your analyzer should be a subclass of the :class:`AnalyzerBase <spack.analyzers.AnalyzerBase>`,
+and you can look at other analyzers in that folder for examples. The guide
+here will tell you about the basic functions needed.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Analyzer Output Directory
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, when you run ``spack analyze`` an analyzer output folder will
+be created in the spec's package directory for results:
```

Comment:
I think I was the one who suggested this location, but the more I think about it the more I think we want to move it.  We haven't always been good at this, but going forward we need to assume that the install tree may be read-only for pretty much any command but `install`.  e.g., in the Axom installations we've talked about analyzing, you don't actually have write access to `$prefix/.spack`, but you'll want to analyze those directories nonetheless.  We'll need a way for you to do that.

Look at `spack test` for an example of how we do it for that command -- the results go in `~/.spack`, which is really the only place we can assume is writable by the current user:

```console
(spackle):cmd> spack test results
==> No test suites with results to report
(spackle):cmd> spack test run libxml2
==> Spack test cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg
==> Testing package libxml2-2.9.10-6zobott
==> Testing package libxml2-2.9.10-dqzlvh6
(spackle):cmd> spack test results    
==> Results for test suite 'cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg':
==>   libxml2-2.9.10-6zobott PASSED
==>   libxml2-2.9.10-dqzlvh6 PASSED
(spackle):cmd> ls ~/.spack                        
cache/  config.yaml  darwin/  mirrors.yaml  test/
(spackle):cmd> ls ~/.spack/test 
compilers.yaml  cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg/
(spackle):cmd> ls ~/.spack/test/cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg 
libxml2-2.9.10-6zobott-test-out.txt*  repo/        test_suite.lock
libxml2-2.9.10-dqzlvh6-test-out.txt*  results.txt
```

I think analyzer results should probably go in there as well.  I am not sure they should be stored the same way.  Tests are stored in a way that allows you to run them many times, and the hash in the test dir is associated with the "suite" or the set of runs.

I think you want to key spack analyzer output by the same key you use for builds in the DB in spack monitor.

It would be nice to figure out if `spack test` and this can share some code, since they have very similar workflows -- tests are just another thing that can be run on already-installed packages.  Analysis isn't so different in that sense.

Reply:
Hmm. For spack monitor the build_id is just an incrementing database index, and the build is unique based on the spack environment (the file that is generated in the install folder):

```
$ cat opt/spack/linux-ubuntu20.04-skylake/gcc-9.3.0/zlib-1.2.11-sl7m27mzkbejtkrajigj3a3m37ygv4u2/.spack/install_environment.json 
{
 "host_os": "ubuntu20.04",
 "platform": "linux",
 "host_target": "skylake",
 "hostname": "vanessa-ThinkPad-T490s",
 "spack_version": "0.16.1-1751-87262723ab",
 "kernel_version": "#74-Ubuntu SMP Wed Jan 27 22:54:38 UTC 2021"
}
```
and then a spec hash. So we definitely have the spec hash, but the above environment isn't really usable as a key. Are you thinking we should try to hash it or something similar? Or create a path there similar to the spec install's path?

Reply:
Yeah I think hashing this would be good.  Then you're guaranteed that whatever you add to the JSON there, you'll still be able to generate unique ids.  Although that doesn't contain the full hash.  Is it Full hash + environment?

One nitpick here: we should probably include the full commit in the spack version, for completeness.  Easiest way to do that might be to tweak `spack --version` to not truncate it.  I don't see anything terrible about doing that, though it'll make the output longer.

Reply:
Hmm, so my concern here is that we want a user to be able to easily match an install to an analyzer directory, and if it's a random hash it would be meaningless. The uniqueness (at least for spack monitor) is the spec full hash plus everything in that environment, which includes the spack version.

I'll take a look at if I can figure out how to include the full commit.

Reply:
I'm testing git describe and I can't figure out how to get the full commit - when I add `--long` there is no result. I might need help with this one.